### PR TITLE
Introducing RSX debugger entry (main debugger)

### PR DESCRIPTION
**Pull request by elad335**
Allows to view complete RSX state. (information about textures, shaders, vertices data, ZCULL and much more)
This string is also logged  upon any RSX crashes.

image

### CODE REVIEW

File: Utilities/Thread.cpp [hunk 1]
```diff
@@ -104,6 +105,21 @@ void fmt_class_string<std::thread::id>::format(std::string& out, u64 arg)
 	out += ss.str();
 }
 
+std::string dump_usefull_thread_info()
```
Comment (Megamouse) [inline Utilities/Thread.cpp:108]:
```suggestion
std::string dump_useful_thread_info()
```

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 2]
```diff
@@ -436,15 +447,22 @@ void debugger_frame::OnSelectUnit()
...
+	const auto render = m_choice_units->currentData().value<rsx::thread*>();
 
-	if (!weak.owner_before(cpu) && !cpu.owner_before(weak))
+	if (!render && !weak.owner_before(cpu) && !cpu.owner_before(weak))
 	{
 		// They match, nothing to do.
 		return;
 	}
 
+	if (render && render == (!Emu.IsStopped() ? m_rsx : nullptr))
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_frame.cpp:458]: Again with the unreadable code...
```suggestion
	if (render && !Emu.IsStopped() && render == m_rsx)
```

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 3]
```diff
@@ -373,20 +380,17 @@ void debugger_frame::UpdateUI()
 			if (cpu->state & cpu_flag::dbg_pause)
 			{
 				m_btn_run->setText(RunString);
-				m_btn_step->setEnabled(true);
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_frame.cpp:376]: Why is this stuff removed all of a sudden

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 4]
```diff
@@ -583,14 +606,12 @@ u64 debugger_frame::EvaluateExpression(const QString& expression)
...
 
-	if (!thread) return 0;
-
 	bool ok = false;
 
 	// Parse expression(or at least used to, was nuked to remove the need for QtJsEngine)
 	const QString fixed_expression = QRegExp("^[A-Fa-f0-9]+$").exactMatch(expression) ? "0x" + expression : expression;
 	const u64 res = static_cast<u64>(fixed_expression.toULong(&ok, 16));
-	return ok ? res : thread->get_pc();
+	return ok ? res : (!Emu.IsStopped() && m_rsx ? +m_rsx->ctrl->get : (thread ? thread->get_pc() : 0));
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_frame.cpp:614]: This is horribly nested ternary...
Just use multiple returns so that one can at least set breakpoints.

File: rpcs3/Emu/RSX/gcm_printing.cpp [hunk 5]
```diff
@@ -1111,28 +1089,47 @@ namespace
 
 	namespace
 	{
+
+		template <u32 opcode, typename = void>
+		struct format_func
+		{
+			//std::type_identity_t<std::string(*)(u32)> operator()()
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/gcm_printing.cpp:1096]: What is this supposed to be

File: rpcs3/Emu/RSX/gcm_printing.cpp [hunk 6]
```diff
@@ -853,13 +854,13 @@ namespace
...
 		{NV3089_IMAGE_IN_OFFSET, "NV3089_IMAGE_IN_OFFSET"},
 		{NV3089_IMAGE_IN, "NV3089_IMAGE_IN"},
-		{GCM_SET_DRIVER_OBJECT, "GCM_SET_DRIVER_OBJECT"},
-		{GCM_DRIVER_QUEUE, "GCM_DRIVER_QUEUE + 0x0"},
-		{GCM_DRIVER_QUEUE+1, "GCM_DRIVER_QUEUE + 0x4"},
-		{GCM_FLIP_HEAD, "GCM_FLIP_HEAD"},
-		{GCM_FLIP_HEAD+1, "GCM_FLIP_HEAD + 0x4"},
-		{GCM_SET_USER_COMMAND, "GCM_SET_USER_COMMAND"},
-		{GCM_FLIP_COMMAND, "GCM_FLIP_COMMAND"},
+		{GCM_SET_DRIVER_OBJECT, "SET_DRIVER_OBJECT"},
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/gcm_printing.cpp:857]: Why are these renamed

Comment (elad335) [reply rpcs3/Emu/RSX/gcm_printing.cpp:857]: GCM_ prefix is redundent for all methods here because it is added elsewhere.

File: rpcs3/Emu/RSX/gcm_printing.cpp [hunk 7]
```diff
@@ -1111,28 +1089,47 @@ namespace
 
 	namespace
 	{
+
+		template <u32 opcode, typename = void>
```
Comment (Nekotekina) [inline rpcs3/Emu/RSX/gcm_printing.cpp:1093]: Please use Camel case for named template parameters.

File: rpcs3/Emu/RSX/rsx_decode.h [hunk 8]
```diff
@@ -73,7 +63,7 @@ struct registers_decoder<NV4097_SET_VIEWPORT_HORIZONTAL>
 
 	static std::string dump(decoded_type &&decoded_values)
```
Comment (Nekotekina) [inline rpcs3/Emu/RSX/rsx_decode.h:64]: Looks like some dump invocations can be moved to cpp file since you are changing them anyway.

Comment (Nekotekina) [reply rpcs3/Emu/RSX/rsx_decode.h:64]: Also using && here is weird, `const decoded_type&` I would pass.

File: rpcs3/rpcs3qt/debugger_frame.cpp (hunk 2 above)
Comment (elad335) [reply rpcs3/rpcs3qt/debugger_frame.cpp]: Fixed.

File: rpcs3/Emu/RSX/rsx_decode.h (hunk 8 above)
Comment (elad335) [reply rpcs3/Emu/RSX/rsx_decode.h]: Define "some".

File: rpcs3/rpcs3qt/debugger_frame.cpp (hunk 3 above)
Comment (elad335) [reply rpcs3/rpcs3qt/debugger_frame.cpp]: Fixed.

File: rpcs3/Emu/RSX/gcm_printing.cpp (hunk 7 above)
Comment (elad335) [reply rpcs3/Emu/RSX/gcm_printing.cpp]: Fixed.

**Comment by elad335** [comment]
Added many fragment textures properties
image

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 9]
```diff
@@ -583,14 +628,15 @@ u64 debugger_frame::EvaluateExpression(const QString& expression)
...
 	bool ok = false;
 
+	const u32 pc = get_rsx() ? +m_rsx->ctrl->get
+		: (thread ? thread->get_pc() : 0);
+
 	// Parse expression(or at least used to, was nuked to remove the need for QtJsEngine)
 	const QString fixed_expression = QRegExp("^[A-Fa-f0-9]+$").exactMatch(expression) ? "0x" + expression : expression;
 	const u64 res = static_cast<u64>(fixed_expression.toULong(&ok, 16));
-	return ok ? res : thread->get_pc();
+	return ok ? res : pc;
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_frame.cpp:639]: Why not write something like:
if (ok) return res;
if (auto thread = get_rsx()) return +thread->ctrl->get;
auto thread = cpu.lock();
if (thread) return thread->get_pc();
return 0;

File: rpcs3/rpcs3qt/debugger_frame.cpp (hunk 4 above)
Comment (elad335) [reply rpcs3/rpcs3qt/debugger_frame.cpp]: There's nothing wrong with nested ternary whne you do them right IMO, written it better.

File: rpcs3/rpcs3qt/debugger_frame.cpp (hunk 9 above)
Comment (elad335) [reply rpcs3/rpcs3qt/debugger_frame.cpp]: Fixed.

**Comment by Megamouse** [comment]
Stop marking conversations as resolved when participants didn't even see the answer yet.
Might as well not answer at all

File: rpcs3/Emu/RSX/rsx_decode.h (hunk 8 above)
Comment (Nekotekina) [reply rpcs3/Emu/RSX/rsx_decode.h]: Full template specializations (`template<>`).

Comment (elad335) [reply rpcs3/Emu/RSX/rsx_decode.h]: In rsx_decode.h, not including specializains by macros and `template <>`
image
I'm afraid you have to be more specific, I won't overwrite so many specializaions.

Comment (Nekotekina) [reply rpcs3/Emu/RSX/rsx_decode.h]: I meant at least those you changed.

Comment (elad335) [reply rpcs3/Emu/RSX/rsx_decode.h]: That's nearly half of them (you can search fmt::format count), it's way too much unrewarding work. I'm sure this file will be refactored several times in the future because it's very inefficient about how it handles it all. (for example decoded_type structure is completely redundent, was there from legacy reasons where it was some sort of union).
I have to disagree doing this in this pull request.

Comment (Nekotekina) [reply rpcs3/Emu/RSX/rsx_decode.h]: IIRC MSVC had a feature to do this automatically.

File: rpcs3/Emu/RSX/RSXThread.cpp [hunk 10]
```diff
@@ -2626,6 +2639,32 @@ namespace rsx
...
+			case NV4097_INVALIDATE_VERTEX_CACHE_FILE:
+			case NV4097_INVALIDATE_ZCULL:
+				continue;
+			default: break;
+			}
+
+			if (auto func = get_raw_printing_function(i))
+			{
+				fmt::append(result, "[%04x] %s\n", i, func(method_registers.registers[i]));
+			}
```
Comment (kd-11) [inline rpcs3/Emu/RSX/RSXThread.cpp:2662]: If there is a chance of the method returning no function, we need a way to identify that from within the debugger.

Comment (elad335) [reply rpcs3/Emu/RSX/RSXThread.cpp:2662]: like say call rsx::get_methid_name for it and print that instead?

**Review by kd-11 — COMMENTED** [review]
RSX interaction seems fine once the fallback behaviour is defined as requested.

File: rpcs3/Emu/RSX/RSXThread.cpp (hunk 10 above)
Comment (kd-11) [reply rpcs3/Emu/RSX/RSXThread.cpp:2662]: yes.

Comment (elad335) [reply rpcs3/Emu/RSX/RSXThread.cpp:2681]: Done.

Comment (kd-11) [reply rpcs3/Emu/RSX/RSXThread.cpp:2681]: My point was more about code integrity, since the code has been tightened up, add an assert here to avoid someone breaking it. We don't need the if() here, just an ensure, viz:
```
auto func = rsx::get_pretty_printing_function(i);
ensure(func);
//do something with func
```
Alternatively:
```
if (func = ..)
else fmt::throw...
```
Since I can see get_pretty_printing_function is not expected to fail, I propose the first version that just asserts.

Comment (elad335) [reply rpcs3/Emu/RSX/RSXThread.cpp]: Fixed.

**Review by kd-11 — APPROVED** [review]

File: rpcs3/Emu/RSX/gcm_enums.cpp [hunk 11]
```diff
@@ -127,372 +127,608 @@ rsx::texture_dimension rsx::to_texture_dimension(u8 in)
...
 		{
-		case comparison_function::never: return "Never";
-		case comparison_function::less: return "Less";
-		case comparison_function::equal: return "Equal";
-		case comparison_function::less_or_equal: return "Less_equal";
-		case comparison_function::greater: return "Greater";
-		case comparison_function::not_equal: return "Not_equal";
-		case comparison_function::greater_or_equal: return "Greater_equal";
-		case comparison_function::always: return "Always";
+		case rsx::comparison_function::never: return "Never";
```
[the remark this answers is not in the document]
Comment (elad335) [reply rpcs3/Emu/RSX/gcm_enums.cpp]: Done.
